crates/red_knot_python_semantic/src/types.rs

```diff
@@ -580,9 +631,137 @@ impl<'db> Type<'db> {
+                        .contains(&Type::string_literal(db, ""))
+                    && target_intersection
+                        .negative(db)
+                        .contains(&Type::LiteralString)
+                    && target_intersection
+                        .negative(db)
+                        .iter()
+                        .all(|neg_ty| !neg_ty.is_subtype_of(db, Type::AlwaysFalsy))
+                {
+                    return true;
+                }
+                // is_subtype_of(~AlwaysTruthy & ~AlwaysFalsy, ~tuple[()] & ~LiteralString)
+                else if self_intersection.negative(db).contains(&Type::AlwaysTruthy)
+                    && self_intersection.negative(db).contains(&Type::AlwaysFalsy)
+                {
+                    let mut target_negative = target_intersection.negative(db).clone();
+                    if target_negative.remove(&Type::LiteralString)
+                        && target_negative.iter().all(|&target_neg_elem| {
+                            self_intersection.negative(db).iter().any(|&self_neg_elem| {
+                                target_neg_elem.is_subtype_of(db, self_neg_elem)
+                            }) || self_intersection.positive(db).iter().any(|&self_pos_elem| {
+                                self_pos_elem.is_disjoint_from(db, target_neg_elem)
+                            })
+                        })
+                    {
+                        return true;
+                    }
+                }
+                // Check that all target positive values are covered in self positive values
+                target_intersection
+                    .positive(db)
+                    .iter()
+                    .all(|&target_pos_elem| {
+                        self_intersection
+                            .positive(db)
+                            .iter()
+                            .any(|&self_pos_elem| self_pos_elem.is_subtype_of(db, target_pos_elem))
+                    })
+                    // Check that all target negative values are excluded in self, either by being
+                    // subtypes of a self negative value or being disjoint from a self positive value.
+                    && target_intersection
+                        .negative(db)
+                        .iter()
+                        .all(|&target_neg_elem| {
+                            // Is target negative value is subtype of a self negative value
+                            self_intersection.negative(db).iter().any(|&self_neg_elem| {
+                                target_neg_elem.is_subtype_of(db, self_neg_elem)
+                            // Is target negative value is disjoint from a self positive value?
+                            }) || self_intersection.positive(db).iter().any(|&self_pos_elem| {
+                                self_pos_elem.is_disjoint_from(db, target_neg_elem)
+                            })
+                        })
+            }
```

Comment:
This seems to effectively just undo https://github.com/astral-sh/ruff/pull/16636, which is unfortunate as that was a nice simplification. Is this really necessary?